### PR TITLE
[AIRFLOW-6786] Add KafkaConsumerHook, KafkaProduerHook and KafkaSensor

### PR DESCRIPTION
**Pull request by dferguson992**
Dear Airflow Maintainers,

Please accept the following PR that

Add the KafkaProducerHook.
Add the KafkaConsumerHook.
Add the KafkaSensor which listens to messages with a specific topic.
Related Issue:
#1311

Issue link: [AIRFLOW-6786](https://issues.apache.org/jira/browse/AIRFLOW-6786)

Make sure to mark the boxes below before creating PR: [x]

- [x]  Description above provides context of the change
- [X] Commit message/PR title starts with [AIRFLOW-NNNN]. AIRFLOW-NNNN = JIRA ID*
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "How to write a good git commit message"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in Contribution Workflow Example.
* For document-only changes commit message can start with [AIRFLOW-XXXX].

Reminder to contributors:

You must add an Apache License header to all new files
Please squash your commits when possible and follow the 7 rules of good Git commits
I am new to the community, I am not sure the files are at the right place or missing anything.

The sensor could be used as the first node of a dag where the second node can be a TriggerDagRunOperator. The messages are polled in a batch and the dag runs are dynamically generated.

Thanks!

Note, as per denied PR #1415, it is important to mention these integrations are not suitable for low-latency/high-throughput/streaming.  For reference, https://github.com/apache/airflow/pull/1415#issuecomment-484429806.

Co-authored-by: Dan Ferguson dferguson992@gmail.com
Co-authored-by: YuanfΞi Zhu

### CODE REVIEW

**Comment by ashb** [comment]
@dferguson992 This PR pre-dates the projects use of Jira so you'll need to sign up for an account at issues.apache.org/jira/ and create one, please.

**Comment by ashb** [comment]
@dferguson992 Good work!. I'll take a look in detail tomorrow. (there will likely be a few rounds of review as this code was quite old)

**Comment by dferguson992** [comment]
@ashb thanks!  i'll be on vacation, back on Tuesday of next week.  Happy to make any adjustments or additions to the code-base once I return.  The code is quite old you're right, but it should be pretty forward compatible with vanilla Kafka.  

I'm not sure if this project shies away from branding, but it may be worthwhile to consider adding support for Confluent Kafka specifically.

**Comment by ashb** [comment]
The first thing I notice is that we now need to rename this:

-  airflow/contrib/hooks/kafka_consumer_hook.py

->
 - airflow/providers/apache/kafka/hooks/kafka_consumer.py

etc. A bit more detail is in https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-21%3A+Changes+in+import+paths, shout if that's not clear.

File: airflow/contrib/hooks/kafka_consumer_hook.py
Comment (ashb) [inline airflow/contrib/hooks/kafka_consumer_hook.py]: timeout should be a parameter (default of 50 is fine)

File: airflow/contrib/hooks/kafka_consumer_hook.py
Comment (ashb) [inline airflow/contrib/hooks/kafka_consumer_hook.py]: `DEFAULT_HOST` etc for constants please

File: airflow/contrib/hooks/kafka_producer_hook.py
Comment (ashb) [inline airflow/contrib/hooks/kafka_producer_hook.py]: Please avoid `**locals`

File: airflow/contrib/hooks/kafka_producer_hook.py
Comment (ashb) [inline airflow/contrib/hooks/kafka_producer_hook.py]: I don't see a `get_pub` method?

File: airflow/contrib/hooks/kafka_producer_hook.py
Comment (ashb) [inline airflow/contrib/hooks/kafka_producer_hook.py]: This close should probably be in a `finally` block.

File: airflow/contrib/sensors/kafka_sensor.py
Comment (ashb) [inline airflow/contrib/sensors/kafka_sensor.py]: Please don't create hooks in init.

Instead store conn_id and topic on self and then do

```
    @cached_property
    def hook(self):
        return KafkaConsumerHook(self.conn_id, self.topic)
```

File: airflow/contrib/sensors/kafka_sensor.py
Comment (ashb) [inline airflow/contrib/sensors/kafka_sensor.py]: This would be nice to have templates, so please add this at the class level:

```
    templated_fields = ( 'topic', )
```

**Review by ashb — CHANGES_REQUESTED** [review]
I also think the "Co-authored-by" in the messages needs email address (we can use github ones) to make it attribute correclty.

**Comment by dferguson992** [comment]
@ashb I've made the recommended changes except for the github email address.  How does one acquire someone else's GH email?

**Comment by ashb** [comment]
@dferguson992 I've force pushed this branch to update the format/add the emails.

I couldn't see where you got "Hanan Shteingart" from as an author- which PR was that?

File: airflow/contrib/hooks/kafka_consumer_hook.py
Comment (kaxil) [inline airflow/contrib/hooks/kafka_consumer_hook.py]: Let's move this call to get_connection, example: https://github.com/apache/airflow/commit/4bec1cc489f5d19daf7450c75c3e8057c9709dbd

Comment (kaxil) [reply airflow/contrib/hooks/kafka_consumer_hook.py]: This is to avoid DB call on Initialisation.

Comment (dferguson992) [reply airflow/contrib/hooks/kafka_consumer_hook.py]: Address in newest commit.

**Comment by dferguson992** [comment]
hey @ashb I'm not sure what else to do for this PR.  Is there anything outstanding before it can be merged?  I fixed as many of the travis CI issues as I could not sure what else to do for this

**Comment by ashb** [comment]
> @dferguson992 I've force pushed this branch to update the format/add the emails.
> 
> I couldn't see where you got "Hanan Shteingart" from as an author- which PR was that?

@dferguson992 Can you answer this Q?

**Comment by dferguson992** [comment]
@ashb I got that co-author by mistake.  on the original PR, that user was the last to comment on it.  but i mistakenly took that user as the PR author, not an interested third-party as I should have.  I've removed that user from the PR notes and the corresponding issue.

**Comment by dferguson992** [comment]
Hey @ashb hope all is well.  I'm not sure what to do to get the Travis CI checks to pass, but the code is functional and includes your suggestions.  Can we move forward with the PR?

**Comment by dmmakita** [comment]
Hi, @dferguson992, I'm having similar needs related to airflow-kafka. May I try to help you with this issue #7407?

**Comment by dferguson992** [comment]
> Hi, @dferguson992, I'm having similar needs related to airflow-kafka. May I try to help you with this issue #7407?

That would be awesome!  But I'm not sure what else needs doing.  If you know more on this, please do go for it.

**Comment by dmmakita** [comment]
Thank you for your response. I will try to do something.

**Comment by gumartinm** [comment]
Looking forward to seeing this change in Airflow.

**Comment by kaxil** [comment]
Can you please rebase to the latest Master?

**Comment by dferguson992** [comment]
@ashb can you please review?   I've made the necessary changes to this PR and there seems to be growing demand for it.  At this time I see there needs to a review by you and some modifications to the travis ci build which is failing for an unknown reason.

**Comment by kaxil** [comment]
> @ashb can you please review? I've made the necessary changes to this PR and there seems to be growing demand for it. At this time I see there needs to a review by you and some modifications to the travis ci build which is failing for an unknown reason.

Can you rebase the PR please to the latest master, we have moved to Github Actions for CI for most of our tests.

**Comment by dferguson992** [comment]
> > @ashb can you please review? I've made the necessary changes to this PR and there seems to be growing demand for it. At this time I see there needs to a review by you and some modifications to the travis ci build which is failing for an unknown reason.
> 
> Can you rebase the PR please to the latest master, we have moved to Github Actions for CI for most of our tests.

I don't have that as an option.  It says merging is blocked.  Should I create a new PR?

**Comment by kaxil** [comment]
> > > @ashb can you please review? I've made the necessary changes to this PR and there seems to be growing demand for it. At this time I see there needs to a review by you and some modifications to the travis ci build which is failing for an unknown reason.
> > 
> > 
> > Can you rebase the PR please to the latest master, we have moved to Github Actions for CI for most of our tests.
> 
> I don't have that as an option. It says merging is blocked. Should I create a new PR?


```
git fetch upstream master
git rebase upstream/master
```

and force push your branch

**Comment by dferguson992** [comment]
> > > > @ashb can you please review? I've made the necessary changes to this PR and there seems to be growing demand for it. At this time I see there needs to a review by you and some modifications to the travis ci build which is failing for an unknown reason.
> > > 
> > > 
> > > Can you rebase the PR please to the latest master, we have moved to Github Actions for CI for most of our tests.
> > 
> > 
> > I don't have that as an option. It says merging is blocked. Should I create a new PR?
> 
> ```
> git fetch upstream master
> git rebase upstream/master
> ```
> 
> and force push your branch

Okay so I've done that and the pre-checks have all gone.  The only issue now is there's an old request for changes from ashb from Feb12th in a stuck state, and I can't seem to remove ashb as a reviewer from this PR

**Comment by dferguson992** [comment]
Okay, my latest error in this build process is this:

> /opt/airflow/docs/_api/airflow/providers/apache/kafka/hooks/kafka_consumer_hook/index.rst:37: WARNING: Unexpected indentation.
> /opt/airflow/docs/_api/airflow/providers/apache/kafka/sensors/kafka_sensor/index.rst:35: WARNING: Field list ends without a blank line; unexpected unindent.
> looking for now-outdated files... none found
> pickling environment... done
> checking consistency... /opt/airflow/docs/_api/airflow/providers/apache/kafka/hooks/index.rst: WARNING: document isn't included in any toctree
> /opt/airflow/docs/_api/airflow/providers/apache/kafka/sensors/index.rst: WARNING: document isn't included in any toctree
> done

Does anyone know how to resolve an issue like this?  @ashb? @kaxil?

**Comment by serkef** [comment]
> Okay, my latest error in this build process is this:
> 
> > /opt/airflow/docs/_api/airflow/providers/apache/kafka/hooks/kafka_consumer_hook/index.rst:37: WARNING: Unexpected indentation.
> > /opt/airflow/docs/_api/airflow/providers/apache/kafka/sensors/kafka_sensor/index.rst:35: WARNING: Field list ends without a blank line; unexpected unindent.
> > looking for now-outdated files... none found
> > pickling environment... done
> > checking consistency... /opt/airflow/docs/_api/airflow/providers/apache/kafka/hooks/index.rst: WARNING: document isn't included in any toctree
> > /opt/airflow/docs/_api/airflow/providers/apache/kafka/sensors/index.rst: WARNING: document isn't included in any toctree
> > done
> 
> Does anyone know how to resolve an issue like this? @ashb? @kaxil?

@dferguson992 for the last 2 warnings, you need to add the new path in sphinx. Check the known issues in [contributing.md](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#documentation)
for the first 2 warnings, I'll add comments in the code where I think this fails. You test locally by building the docs.

File: airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py
Comment (serkef) [inline airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:71]: Maybe you need a breakline before return argument. Also you don't document `timeout_ms`

File: airflow/providers/apache/kafka/hooks/kafka_producer_hook.py
Comment (serkef) [inline airflow/providers/apache/kafka/hooks/kafka_producer_hook.py:64]: Docstrings are empty here

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (serkef) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]: Maybe you need an empty line before the list of parameters

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (serkef) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]: I think you need to put it in the same line, or use indentation

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (serkef) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]: this fits in one line

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (serkef) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]: hm maybe the check is the opposite? if no messages, return false

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (serkef) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py:30]: why not localhost?

File: airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py
Comment (serkef) [inline airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:35]: Is this attribute `_conn` used ?

Comment (dferguson992) [reply airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:35]: yeah if it's set externally by airflow connections it sets the hostname and port number.  if not, the defaults are used.

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (dferguson992) [reply airflow/providers/apache/kafka/sensors/kafka_sensor.py]: ah great catch!  this needs to be there

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]:
```suggestion
        """
        Returns a Kafka Consumer Hook
```

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]:
```suggestion
            self.log.info('Got messages during poking: %s', str(messages))
            return messages
```

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py]:
```suggestion
        self.log.info('Poking topic: %s, using hook: %s', str(self.topic), str(self.hook))
```

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (dferguson992) [reply airflow/providers/apache/kafka/sensors/kafka_sensor.py:30]: I believe when you build kafka locally, many of the apache guideline and walkthroughs recommend going through /etc/hosts and adding kafka1 as an entry, alongside zookeeper innstance etc.

**Comment by dferguson992** [comment]
okay so this one is confusing me.  currently the build is halted during the "backport packages" build step because of the following error: "ModuleNotFoundError: No module named 'kafka'"

What could cause that?

**Comment by ashb** [comment]
@dferguson992 You are importing things from the `kafka` library, but haven't added it as a requirement to setup.py, so you need to do that at least. See also https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#backport-providers-packages

**Comment by dferguson992** [comment]
hey @ashb i think we're good now, just need your approval to merge

**Comment by dferguson992** [comment]
> hey @ashb i think we're good now, just need your approval to merge

never mind.  the build was working earlier, I've honestly no idea why it keeps working and then not working.  the auto-generated documentation is messed up now.  will keep trying to fix.

**Comment by haidaraM** [comment]
Any news on this PR ?

**Comment by dferguson992** [comment]
Working on it @haidaraM, the build process is taking longer than I expected

**Comment by mhbrown** [comment]
Typo in the title: "KafkaProduerHook" --> "KafkaProducerHook"

**Comment by rhps** [comment]
Hi Guys, anyone can help with these issues? @ashb

**Comment by mik-laj** [comment]
I think there is something wrong with this change because I can see too many files. My suggestion is to copy all the relevant files to a separate directory and then create the entire branch from scratch (git reset --hard apache/master). 

When there are fewer files on the branch, I will be very happy to review this change.

CC: @turbaszek @potiuk Can you help this change if I was unavailable? I think it is important from the Apache Foundation's perspective to ensure good integration between the different products it develops.

**Comment by mik-laj** [comment]
@dferguson992  I would like to apologize that you have to wait so long for this change to be reviewed. Please post the progress of this change on the #development channel so that any active contributor can see that you are pending a review. If you do not get any response within 2 days, please retry the messages as we may be busy at the moment.  I would like this change to be completed quickly so that users can benefit from the effects of your work.

**Comment by haidaraM** [comment]
Agreed with @mik-laj . Looks like the PR got lot of unrelated changes somehow.

**Comment by kaxil** [comment]
I have rebased to the latest master and force-pushed to this PR.

🤞 that CI passes (There were many changes -- hopefully I haven't left anything behind)

File: airflow/providers/apache/kafka/example_dags/example_kafka_dag.py
Comment (kaxil) [inline airflow/providers/apache/kafka/example_dags/example_kafka_dag.py:47]: Can we add some example values over here, please

File: airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py
Comment (kaxil) [inline airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:77]: Is this needed ? @dferguson992

File: airflow/providers/apache/kafka/hooks/kafka_producer_hook.py
Comment (kaxil) [inline airflow/providers/apache/kafka/hooks/kafka_producer_hook.py:42]:
```suggestion
      Returns a Kafka Producer
```

File: airflow/providers/apache/kafka/hooks/kafka_producer_hook.py
Comment (kaxil) [inline airflow/providers/apache/kafka/hooks/kafka_producer_hook.py:45]:
```suggestion
        :returns: A Kafka Producer object.
```

File: airflow/providers/apache/kafka/hooks/kafka_producer_hook.py
Comment (kaxil) [inline airflow/providers/apache/kafka/hooks/kafka_producer_hook.py:53]:
```suggestion
            self.server = f"{host}:{port}"
```

File: airflow/providers/apache/kafka/hooks/kafka_producer_hook.py
Comment (kaxil) [inline airflow/providers/apache/kafka/hooks/kafka_producer_hook.py:81]:
```suggestion
     A pretty version of the connection string.
```

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py:35]:
```suggestion
    templated_fields = ('topic', 'host', 'port', )
```

They can fit in one line :)

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py:47]: Let's add some descriptions for each of the docstrings

File: airflow/providers/apache/kafka/sensors/kafka_sensor.py
Comment (kaxil) [inline airflow/providers/apache/kafka/sensors/kafka_sensor.py:63]:
```suggestion
     Checks to see if messages exist on this topic/partition.
```

**Comment by kaxil** [comment]
I have also squashed your 121 commits to 1 :)

**Review by kaxil — COMMENTED** [review]
Tests needs to be added too. This PR in it's current state has no tests

File: airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py
Comment (turbaszek) [inline airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:32]: Would you mind adding type hints? There's already ongoing effort to improve mypy coverage #9708

File: airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py
Comment (turbaszek) [inline airflow/providers/apache/kafka/hooks/kafka_consumer_hook.py:92]: Not sure if this will be ever used

**Comment by turbaszek** [comment]
What do you think about having single `KafkaHook` that implements both consumer and producer methods? It seems to be common approach in Airflow to try to have single hook for an external service.

**Comment by danfranks** [comment]
Can't wait to see this implemented.  We use kafka as our data backbone company-wide, and coordinating between different parts of the org is done through kafka topics.  Producing workflow status completion writes to kafka topics that are consumed by applications, triggering automated report generation, and delivered to external clients.  Then also tracking which clients get what data for CCPA and GDPR comms.

**Comment by turbaszek** [comment]
@dferguson992 do you need any help?

**Comment by dferguson992** [comment]
I do need a lot of help.  I've never successfully gotten these build checks to pass and I don't know what to do about it anymore.  I'm too busy with my day job to really dive into what's wrong, and when I have spent time digging into it I can't understand how all of the pieces work together.  Definitely want some help here, there's only three classes being added and I don't know what the process is for getting this merged back in or how to build integrated tests, updating documentation, etc.etc.

**Comment by potiuk** [comment]
Hey @dferguson992 - sad to see this is difficult. Unfortunately, we have quite some expectations when it comes to code contributions. We keep on discussing it, how big of an obstacle it is for new contributors, but as a community, we need to make sure that we can take over the code and maintain it further when it gets contributed. This often means documentation and code linting, and following the same practices for all parts of the code. So apologies if it's a bit involved - it is so for a reason.

We are happy to help if you would like to learn how to do it. It's OK if takes weeks or months to contribute even small change, so if you have just a little time to spare a week, it can be done in baby steps. We understand the whole process can be a little overwhelming, but if you approach it one thing at a time this might work. And we encourage you to do so - the Kafka hooks/sensors seem to be a really cool addition.

I am happy to provide some guidance. I think what is really important to start with is that you learn how to rebase your changes on top of master. You will have to do it as the first thing every time you come back to working on this, and it's easier than you think. We have some helpful guide on it here: https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#how-to-rebase-pr. And making sure the documentation is updated after the rebase, might be the first thing that you can fix without much of learning I think. Maybe start with that and let's do it together one thing at a time? What do you think?

Feel free to reach out here or in slack if you need some help. I hope we can make it together :)

**Comment by dferguson992** [comment]
Thank you so much @potiuk!  I'd love to join the slack channel for this.  I think that will help me the most since I seldom have time to sit and grind away at the issues I'm encountering.  Getting on the slack would help me to post my issues and look at the results later.  I'll start revisiting this PR and will post my issues in the slack.

**Comment by potiuk** [comment]
> Thank you so much @potiuk! I'd love to join the slack channel for this. I think that will help me the most since I seldom have time to sit and grind away at the issues I'm encountering. Getting on the slack would help me to post my issues and look at the results later. I'll start revisiting this PR and will post my issues in the slack.

Cool. The best places to ask are #development for development questions and #airflow-how-to-pr . Or #newbie-questions if you are shy :)

**Comment by turbaszek** [comment]
Not sure if this CI error about cyclic imports is related... @potiuk

**Comment by knackjason** [comment]
Did this ever get merged in? I'd be interested in having officially supported Kafka hooks.

**Comment by potiuk** [comment]
Does not look like

**Comment by potiuk** [comment]
But there is this PR in progress: https://github.com/apache/airflow/pull/12388

**Comment by knackjason** [comment]
> But there is this PR in progress: #12388

Ah, I didn't realize there was a newer PR. Thanks, @potiuk!

**Pull request closed without merging**
